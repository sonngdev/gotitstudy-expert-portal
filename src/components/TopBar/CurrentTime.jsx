import React, { useState, useEffect } from 'react';
import { Icon } from '@gotitinc/design-system';
import { getCstTime } from '../../helpers';

export function CurrentTime() {
  const [time, setTime] = useState(getCstTime());

  useEffect(() => {
    const interval = setInterval(() => setTime(getCstTime()), 1000);
    return () => clearInterval(interval);
  }, []);

  return (
    <div className="current-time u-text200 u-paddingBottomTiny">
      <Icon name="calendar" size="extraSmall" className="u-marginRightTiny" />
      {time}
    </div>
  )
}
