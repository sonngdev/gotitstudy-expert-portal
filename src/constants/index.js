export const DEFAULT_TIMEZONE = 'America/Chicago';
/**
 * `z` is deprecated in moment, but not in moment-timezone.
 *
 * We need to use `z` in our format, but we cannot test the output by trying
 * to parse it back (using moment) to see if it is valid. This is because to
 * moment (at least in newer versions) it is not.
 *
 */
export const LONG_TIME_FORMAT = 'MMM D, YYYY HH:mm:ss z';
/**
 * A session is 5 minutes (300,000 miliseconds) long.
 */
export const SESSION_DURATION = 300000;
