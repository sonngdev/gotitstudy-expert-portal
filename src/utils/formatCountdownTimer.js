export function formatCountdownTimer(milisecond) {
  const minute = Math.floor(milisecond / 1000 / 60);
  const second = Math.floor(milisecond / 1000) - 60 * minute;

  return [minute, second].map((n) => String(n).padStart(2, '0')).join(':');
}
