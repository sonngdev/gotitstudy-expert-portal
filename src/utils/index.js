export { getCurrentTime, formatCountdownTimer, formatCurrentTime } from './time';
export { useChatMessagesList, useChatMessagesAdd } from './useChatMessages';
export { useScrollToBottom } from './useScrollToBottom';
export { useTinyMceEditor } from './useTinyMceEditor';
