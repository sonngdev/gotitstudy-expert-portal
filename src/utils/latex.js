/* eslint-disable prefer-arrow-callback */

tinymce.PluginManager.add('latex', function addLatexPlugin(editor) {
  /**
  |--------------------------------------------------
  | VARIABLES
  |--------------------------------------------------
  */
  const latexClass = 'Latexformula';

  const latexPreviewId = `${latexClass}-preview`;


  /**
  |--------------------------------------------------
  | HELPERS
  |--------------------------------------------------
  */
  const isLatexFormula = (node) => [...node.classList].includes(latexClass);

  const getExpression = (dialog) => dialog.getData().expression.trim();

  const getSelectedExpression = () => {
    const { selection } = editor;
    const node = selection && selection.getNode();
    return node && isLatexFormula(node) ? node.alt : '';
  };

  const getFormulaSrc = (expression) => (
    /**
     * LaTeX Math as SVG image:
     * https://math.now.sh/home
     */
    `https://math.now.sh?from=${encodeURIComponent(expression)}`
  );

  const setPreview = (content) => {
    document.getElementById(latexPreviewId).innerHTML = content;
  };


  /**
  |--------------------------------------------------
  | DIALOG
  |--------------------------------------------------
  */
  const openDialog = () => editor.windowManager.open({

    title: 'LaTeX expression',

    body: {
      type: 'panel',
      items: [
        { type: 'textarea', name: 'expression', label: 'Expression' },
        { type: 'button', name: 'preview', text: 'Preview' },
        { type: 'htmlpanel', html: `<div id="${latexPreviewId}"></div>` },
      ],
    },

    buttons: [
      { type: 'cancel', text: 'Cancel' },
      { type: 'submit', text: 'Insert', primary: true },
    ],

    initialData: {
      expression: getSelectedExpression(),
    },

    onAction(dialog, details) {
      if (details.name !== 'preview') return;

      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img src="${src}" alt="${expression}" role="math" />`;
      setPreview(content);
    },

    onSubmit(dialog) {
      const expression = getExpression(dialog);
      if (!expression) return;

      const src = getFormulaSrc(expression);
      const content = `<img class="${latexClass}" src="${src}" alt="${expression}" role="math"/>`;
      editor.insertContent(content);

      dialog.close();
    },
  });


  /**
  |--------------------------------------------------
  | REGISTER
  |--------------------------------------------------
  */
  editor.ui.registry.addButton('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.ui.registry.addMenuItem('latex', {
    text: 'LaTeX',
    onAction: openDialog,
  });

  editor.on('DblClick', ({ target }) => {
    if (isLatexFormula(target)) openDialog();
  });

  return {
    getMetadata: () => ({ name: 'LaTeX plugin for tinymce 5' }),
  };
});
