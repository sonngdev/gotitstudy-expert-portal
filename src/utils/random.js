/**
 * For trivial randomness, this solution is sufficient
 * https://stackoverflow.com/questions/1349404/generate-random-string-characters-in-javascript
 */
export function weakRandomString(len) {
  if (len >= 10) throw new Error('String length must be smaller than 10');
  return Math.random().toString(36).substring(2, 2 + len);
}
