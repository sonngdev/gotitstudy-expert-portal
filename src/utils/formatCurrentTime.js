import moment from 'moment-timezone';
import { DEFAULT_TIMEZONE, CURRENT_TIME_FORMAT } from '../constants';

export function formatCurrentTime() {
  return moment().tz(DEFAULT_TIMEZONE).format(CURRENT_TIME_FORMAT);
}
