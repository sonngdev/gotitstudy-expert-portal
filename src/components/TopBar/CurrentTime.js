import React, { useState, useEffect } from 'react';
import { Icon } from '@gotitinc/design-system';
import { getCurrentTime, formatCurrentTime } from '../../utils';

export function CurrentTime() {
  const [time, setTime] = useState(getCurrentTime());

  useEffect(() => {
    const interval = setInterval(() => setTime(getCurrentTime()), 1000);
    return () => clearInterval(interval);
  }, []);

  return (
    <div className="current-time u-text200 u-paddingBottomTiny">
      <Icon name="calendar" size="extraSmall" className="u-marginRightTiny" />
      {formatCurrentTime(time)}
    </div>
  );
}
