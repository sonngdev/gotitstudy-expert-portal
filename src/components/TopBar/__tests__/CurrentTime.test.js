import React from 'react';
import moment from 'moment-timezone';
import { shallow } from 'enzyme';
import { getCurrentTime } from '../../../utils';
import { DEFAULT_TIMEZONE } from '../../../constants';
import { CurrentTime } from '../CurrentTime';

jest.mock('../../../utils', () => ({
  __esModule: true,
  /**
   * Mock only part of a module, keep original implementation of
   * other functions.
   * https://stackoverflow.com/a/53402206/9744063
   *
   * Another way is documented in SessionTimer.test.js
   */
  ...(jest.requireActual('../../../utils')),
  getCurrentTime: jest.fn(),
}));

describe('<CurrentTime />', () => {
  let wrapper;

  beforeEach(() => {
    /**
     * Use `mockReset` to clear mock return values
     */
    getCurrentTime.mockReset();
    getCurrentTime
      .mockReturnValueOnce(moment(1581473066000).tz(DEFAULT_TIMEZONE))
      .mockReturnValueOnce(moment(1581473067000).tz(DEFAULT_TIMEZONE));
    wrapper = shallow(<CurrentTime />);
  });

  it('renders', () => {
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('contains calendar icon', () => {
    const calendarIcon = wrapper.find('Icon');
    expect(calendarIcon.exists()).toBe(true);
    expect(calendarIcon.prop('name')).toBe('calendar');
  });

  it.skip('shows current time', () => {
    jest.useFakeTimers();

    expect(wrapper.text()).toBe('Feb 11, 2020 20:04:26 CST');

    jest.advanceTimersByTime(1000);
    expect(wrapper.text()).toBe('Feb 11, 2020 20:04:27 CST');

    jest.useRealTimers();
  });
});
