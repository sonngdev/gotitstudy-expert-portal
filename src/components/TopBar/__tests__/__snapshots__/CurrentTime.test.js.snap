// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrentTime /> renders 1`] = `
"<div className=\\"current-time u-text200 u-paddingBottomTiny\\">
  <Icon name=\\"calendar\\" size=\\"extraSmall\\" className=\\"u-marginRightTiny\\" />
  Feb 11, 2020 20:04:26 CST
</div>"
`;
